### PR TITLE
0048/transect transport

### PR DESCRIPTION
**Pull request by anwiseNOCL**
- Transect class has been created which will allows the user to:

1. Define a transect between two given (lat,lon) points
2. Calculate the 4d velocities across the transect
3. Calculate the 3d depth integrated volume transport across the transect
4. Calculate the depths at the u and v points along the transect (this currently uses depth_t_0 as an example and will be changed once the new NEMO class structure #105  and depth calculation #109  code is complete)
5. Allows the user to plot the transport and normal velocities.

- Depth_0 calculations added to DOMAIN
- A tweak has been made to the DOMAIN method that calculates the transect indices. This was necessary to ensure the indices are 'regularly' spaced. NOTE that this broke 3a and 4a in the unit testing, which have now been updated.
- The test data file are located in /projesctsa/COAsT/nemo_data_T(U)(V)_grid.nc and need to be put into the example_files folder to run the unit_test

### CODE REVIEW

**Comment by davbyr** [comment]
@anwiseNOCL I can't get the example files from /projectsa/ to test this. Could you change the permission?

```
-rw-rw----. 1 anwise visitors  84314971 Jul  6 14:51 nemo_data_T_grid.nc
-rw-rw----. 1 anwise visitors  26119626 Jul  6 14:51 nemo_data_U_grid.nc
-rw-rw----. 1 anwise visitors  26623241 Jul  6 14:52 nemo_data_V_grid.nc
```

File: coast/TRANSECT.py [hunk 1]
```diff
@@ -1,18 +1,313 @@
...
+        # point A should be of lower latitude than point B
+        if abs(point_B[0]) < abs(point_A[0]):
+            self.point_A = point_B
+            self.point_B = point_A
+        else:
+            self.point_A = point_A
+            self.point_B = point_B
+            
+        # Get points on transect
+        tran_y, tran_x = self.get_transect_indices( domain, self.point_A, self.point_B, "f" )
```
Comment (jpolton) [inline coast/TRANSECT.py:50]: grid_ref now has syntax "{tuvf}_grid" to make it more grep-able.

Comment (jpolton) [inline coast/TRANSECT.py:50]: tran_y and tran_x are indices, should they be labelled as such: tran_y_ind, tran_x_ind ? Or call them "itran_x" and "itran_y" ?

File: coast/TRANSECT.py [hunk 2]
```diff
@@ -1,18 +1,313 @@
...
+                
+        # indices along the transect        
+        self.y_idx = tran_y
```
Comment (jpolton) [inline coast/TRANSECT.py:53]: "idx" could look like "i_delta x". Perhaps "_ind" is cleaner.

File: coast/TRANSECT.py [hunk 3]
```diff
@@ -1,18 +1,313 @@
...
+        self.x_idx = tran_x
+        self.len = len(tran_y)
+        self.data = xr.Dataset()
+        self.normal_velocities = None
+        self.depth_integrated_transport_across_AB = None # (time,transect_segment_index)
+        
+        # dataset along the transect
+        da_tran_y = xr.DataArray( tran_y, dims=['s_dim_f_grid'])
```
Comment (jpolton) [inline coast/TRANSECT.py:61]: "da_tran_y": Again, these are a da of indices, add _ind to the name?

File: coast/TRANSECT.py [hunk 4]
```diff
@@ -1,18 +1,313 @@
...
+        
+        velocity = np.ma.zeros(np.shape(self.data_U.vozocrtx))
+        vol_transport = np.ma.zeros(np.shape(self.data_U.vozocrtx))
+        depth_integrated_transport = np.ma.zeros( np.shape(self.data_U.vozocrtx[:,0,:] )) 
+        depth = np.ma.zeros(np.shape(self.data_U.vozocrtx))
+        
+        dy = np.diff(self.y_idx)
+        dx = np.diff(self.x_idx)
+        # Loop through each point along the transact
+        for idx in np.arange(0, self.len-1):
```
Comment (jpolton) [inline coast/TRANSECT.py:139]: idx --> ind ?

File: unit_testing/unit_test.py [hunk 5]
```diff
@@ -163,20 +163,71 @@
...
+length_ref = 37
 
 
 if (xt == xt_ref) and (yt == yt_ref) and (length_of_line == length_ref):
     print(str(sec) + chr(subsec) + " OK - NEMO domain transect indices extracted")
 else:
     print(str(sec) + chr(subsec) + " X - Issue with indices extraction from NEMO domain transect")
 
+#-----------------------------------------------------------------------------#
+# ( 3b ) Transport velocity and depth calculations                            #
```
Comment (jpolton) [inline unit_testing/unit_test.py:181]: 3b and 3c do not appear in unit_test_contents.txt

File: unit_testing/unit_test.py [hunk 6]
```diff
@@ -163,20 +163,71 @@
...
+tran.data_V = tran.data_V.assign(e3v=e3v)
+
+output = tran.transport_across_AB()
+# Check the calculations are as expected
+if np.isclose(tran.data.depth_integrated_transport_across_AB.sum(), -49.19533238588342)  \
+        and np.isclose(tran.data.depth.sum(), 16116315.485839844) \
+        and np.isclose(np.nansum(tran.data.normal_velocities.values), -253.6484375): 
+    print(str(sec) + chr(subsec) + " OK - TRANSECT transport velocities good")
+else:
+    print(str(sec) + chr(subsec) + " X - TRANSECT transport velocities no good")
```
Comment (jpolton) [inline unit_testing/unit_test.py:211]: no --> not

File: unit_testing/unit_test.py [hunk 7]
```diff
@@ -163,20 +163,71 @@
...
+subsec = subsec+2
+
+plt.close('all')
+try:
+    plot_dict = {'fig_size':(5,3), 'title':'Normal velocities'}
+    fig,ax = tran.plot_normal_velocity(time='2010-01-02',cmap="seismic",plot_info=plot_dict,smoothing_window=2)
+    fig.savefig(dn_fig + 'transect_velocities.png')
+    plot_dict = {'fig_size':(5,3), 'title':'Transport across AB'}
+    tran.plot_depth_integrated_transport(time=0, plot_info=plot_dict, smoothing_window=2)
+    fig.savefig(dn_fig + 'transect_transpor.png')
```
Comment (jpolton) [inline unit_testing/unit_test.py:226]: transpor.png --> transport.png

**Comment by jpolton** [comment]
I'm more sure the gitignore statement is working properly:
COAsT/example_files/*.nc --> example_files/*.nc

**Pull request merged**